Documentation/replacing-component.md

```diff
@@ -0,0 +1,266 @@
+
+```swift
+protocol Component {
+    associatedtype Content: Component = Never
+    associatedtype Result: ResponseEncodable = Never
+    
+    @ComponentBuilder
+    var content: Content { get }
+    
+    func handle() -> Response
+}
+```
+
+
+- `content`: used to provide further DSL components
+- `handle()`: used to specify how the component should respond to incoming requests
+
+The issue here is that these two members are, in effect, mutually exclusive: if a component implements both `content` and `handle()`, we currently ignore its `handle()` function and only look at `content`.
+
+
+
+
+If we think of the `WebService`, as defined via the DSL, as a tree data structure, we can come up with the following mapping:
+
+- root node: the entry point into the `WebService`
+- internal nodes (ie non-leaf nodes): intermediate components used for grouping, which eventually lead to one or more endpoints
+- leaves: endpoints, which expose functionality via their `handle()` function
+
+
+The issue with the current definition of the `Component` protocol is that it is trying to cover all of these three cases at the same time, with the same interface.
+
+
+One of the additional side effects of switching to a more fine-grained protocol structure is that we can use the type system to express some of the requirements a web service must satisfy. For example, we should be able to define the protocols/types/builder in a way that ill-formed web services will cause a compile-time error.
+
+
+
+
+### Issues caused by the current implementation
+
+
+#### Invalid Components
+
+Both of these components currently compile, despite not being valid components:
+
+```swift
+// empty component which implements neither `content` nor `handle()`
+struct A: Component {}
+
+// component which implements both `content` and `handle()`
+struct B: Component {
+    var content: some Component { ... }
+    func handle() -> Response { ... }
+}
+```
+
+More generally speaking, fully half of the four possible combinations of conforming to the `Component` protocol will result in invalid components:
+
+<!--- Implement neither `content` nor `handle()`: ❌
+- Implement only `content`: ✅
+- Implement only `handle()`: ✅
+- Implement both `content` and `handle()`: ❌-->
+
+
+|            | `content` | `handle()` | component valid? |
+| :--------- | :-------: | :--------: | :--------------: |
+| implements |  no       | no         | ❌       |
+| implements |  yes      | no         | ✅               |
+| implements |  no       | yes        | ✅               |
+| implements |  yes      | yes        | ❌               |
+
+
+
+
+#### Type inference
+
+Because the current implementation defaults both `Content` and `Response` to `Never`, you can end up in situations where a seemingly properly defined component is in fact ill-formed.
+
+
+Consider the following example:
+
+```swift
+struct User: Codable {
+    let id: ID
+    let name: String
+    let age: Int
+}
+
+struct GetUser: Component {
```

Comment:
This is an interesting edge case that I have not considered yet. I suspect this would also happen in this implementation for the `Handler`: https://github.com/Apodini/Apodini/pull/26#discussion_r538396896 ... so I suspect we would need to remove the `Never` default value?

Reply:
This specific example would in fact work as expected now, since https://github.com/Apodini/Apodini/commit/0811646d1c87cb51b4817402a2228d2968e201fb.

It'd still break if `User` isn't `Codable` anymore.

Yes, removing the default type would fix this